torch/distributed/_tensor/_collective_utils.py

```diff
@@ -167,65 +169,85 @@ def spec_to_bytes(spec: "placement_types.DTensorSpec") -> int:
     return spec.tensor_meta.dtype.itemsize * math.prod(spec.shape)
 
 
-def get_bandwidth_factor(mesh: DeviceMesh) -> List[float]:
-    # generate bandwidth factor for intra-host/inter-host communication pattern
-    factors = [1.0] * mesh.ndim
-    num_devices_per_host = _mesh_resources.num_devices_per_host(mesh.device_type)
-
-    num_devices = 1
-    for mesh_dim in reversed(range(mesh.ndim)):
-        num_devices *= mesh.size(mesh_dim)
-        if num_devices <= num_devices_per_host:
-            # magic number for intra-host communication bandwidth factor
-            # TODO: see if we need to tweak this or offer a way for user
-            # to specify the bandwidths
-            factors[mesh_dim] = 0.2
-
-    return factors
-
-
-def allgather_cost(num_bytes: float, mesh: DeviceMesh, mesh_dim: int) -> float:
-    num_devices_on_mesh_dim = mesh.size(mesh_dim)
-    bandwidth_factor = get_bandwidth_factor(mesh)[mesh_dim]
-    # constant latency factor + bandwidth cost
-    return (
-        1
-        + bandwidth_factor
-        * num_bytes
-        * (num_devices_on_mesh_dim - 1)
-        / num_devices_on_mesh_dim
-    )
+@dataclass
+class MeshTopoInfo:
+    """
+    Mesh information for collective cost estimation
+    """
 
+    mesh: DeviceMesh
+    mesh_dim_devices: List[int]
+    mesh_dim_bandwidth: List[float]
+    mesh_dim_latency: List[float]
+
+    @staticmethod
+    @lru_cache(None)
+    def build_from_mesh(mesh: DeviceMesh) -> "MeshTopoInfo":
+        # Generate mesh topology info for intra-host/inter-host communication pattern
+        # Note that we made bunch of assumptions for simplicity:
+        # 1. we assume the mesh is homogeneous, and it's gpu/nccl model
+        # 2. we assume gpu arch is Ampere or Hopper
+        # 3. we assume collectives are all ring base algo for now
+        num_devices_per_host = _mesh_resources.num_devices_per_host(mesh.device_type)
+        # the base bw number (intra-node), GB/s
+        base_bw = 87.7
```

Comment:
Where did we get this intra-node bandwidth number?

Reply:
mostly from here https://github.com/pytorch/pytorch/blob/main/torch/_inductor/comm_analysis.py#L156

Reply:
Since these numbers are per channel and NCCL uses multiple channels, would this base bandwidth number be low? E.g. normally intra-node bandwidth on Ampere is ~300 GB/s I thought.

Reply:
Yeah thanks for catching this! I'm trying to figure out what exactly is the per-GPU bandwidth to other GPUs, in the case of NVLink. 300 GB/s means uni-directional p2p bandwidth, is this counting the channels already?

There's some bus bandwidth calculation where it counts for # channels = 2, and have some bandwith refinements, if we take this formula down for intra-node then it's around 2 * 300 /3 = 200 https://github.com/pytorch/pytorch/blob/main/torch/_inductor/comm_analysis.py#L223-L227 I am not really sure it should be 200 here, as we are really caculating the bus bandwidth from an algorithm bandwidth. So for nvlink if the algo bwdwidth should be 300 then the base_bw here should be 300 instead I guess?

Reply:
🤔 maybe the comm analysis pass bandwidth estimation is kinda wrong too, as it would almost always take the smaller one, `llMaxBw` which is 87.6